tests/cli/test_rasa_train.py

```diff
@@ -194,6 +199,75 @@ def test_train_force(run_in_simple_project_with_model: Callable[..., RunResult])
     assert len(files) == 2
 
 
+def test_train_dry_run(run_in_simple_project_with_model: Callable[..., RunResult]):
+    temp_dir = os.getcwd()
+
+    assert os.path.exists(os.path.join(temp_dir, "models"))
+    files = rasa.shared.utils.io.list_files(os.path.join(temp_dir, "models"))
+    assert len(files) == 1
+
+    output = run_in_simple_project_with_model("train", "--dry-run")
+    printed_output = set(output.outlines)
+
+    assert [s for s in printed_output if "No training required." in s]
+    assert output.ret == 0
+
+
+def test_train_dry_run_failure(
+    run_in_simple_project_with_model: Callable[..., RunResult]
+):
+    temp_dir = os.getcwd()
+
+    domain = (
+        "version: '2.0'\n"
+        "session_config:\n"
+        "  session_expiration_time: 60\n"
+        "  carry_over_slots_to_new_session: true\n"
+        "actions:\n"
+        "- utter_greet\n"
+        "- utter_cheer_up"
+    )
+
+    with open(os.path.join(temp_dir, "domain.yml"), "w") as f:
+        f.write(domain)
+
+    output = run_in_simple_project_with_model("train", "--dry-run")
+    printed_output = set(output.outlines)
+
+    assert not any([s for s in printed_output if "No training required." in s])
+    assert (
+        (
+            output.ret & CODE_CORE_NEEDS_TO_BE_RETRAINED
```

Comment:
this is pretty cool. Is this a conventional way of handling shell exit codes?

Reply:
Well, afaik the standard exit code is 0 for success and any number from 1 to 255 for anything else. 
I thought that it might be a good idea to just use bit masks to encode all the information we need because we have 4 booleans in `FingerprintComparisonResult` (which gives us 2^4 possible results).